File: scripts/validate-browser-cohort.mjs
import { createHash } from "node:crypto";
import { lstat, readdir, readFile, readlink } from "node:fs/promises";
import { join, relative, sep } from "node:path";

const ROOT =
	"/Users/pranav/Library/Caches/ms-playwright/chromium-1234/chrome-mac-arm64/Google Chrome for Testing.app";
const LAUNCHER = join(ROOT, "Contents/MacOS/Google Chrome for Testing");
const FRAMEWORK = join(
	ROOT,
	"Contents/Frameworks/Google Chrome for Testing Framework.framework/Versions/151.0.7922.34/Google Chrome for Testing Framework",
);
const EXPECTED = Object.freeze({
	manifestBytes: 62_239,
	manifestHash:
		"25995bc88bf20b6de47b46eb3571b250989846a797c0b8924b8794627b6175fc",
	files: 326,
	links: 5,
	totalBytes: 372_002_382,
	launcherHash:
		"a596b1cfc6353e987fcec8d71a23a28cd6a9e7a6b4e20b908e4c4fcffe51158e",
	frameworkBytes: 237_813_488,
	frameworkHash:
		"269114cf695f1c50b54e0816a1442e41dc468d28672e2dedc2036105fb5a8dbe",
});

const sha256 = (bytes) => createHash("sha256").update(bytes).digest("hex");
const fail = (message) => {
	throw new Error(`browser cohort violation: ${message}`);
};
const safe = (value) =>
	!/[\t\n\r]/.test(value) && !value.split("/").includes("..");
const mode = (stat) => (stat.mode & 0o7777).toString(8).padStart(4, "0");

const entries = [];
async function walk(path) {
	for (const name of await readdir(path)) {
		const absolute = join(path, name);
		const stat = await lstat(absolute);
		const rel = relative(ROOT, absolute).split(sep).join("/");
		if (!safe(rel)) fail(`unsafe relative path ${rel}`);
		if (stat.isDirectory()) await walk(absolute);
		else if (stat.isFile()) entries.push({ kind: "F", absolute, rel, stat });
		else if (stat.isSymbolicLink()) {
			const target = await readlink(absolute);
			if (!safe(target)) fail(`unsafe symlink target ${rel}`);
			entries.push({ kind: "L", absolute, rel, stat, target });
		} else fail(`unsupported object ${rel}`);
	}
}

await walk(ROOT);
entries.sort((a, b) => Buffer.compare(Buffer.from(a.rel), Buffer.from(b.rel)));
let files = 0;
let links = 0;
let totalBytes = 0;
const lines = [];
for (const entry of entries) {
	if (entry.kind === "F") {
		const bytes = await readFile(entry.absolute);
		files += 1;
		totalBytes += bytes.byteLength;
		lines.push(
			`F\t${entry.rel}\t${mode(entry.stat)}\t${bytes.byteLength}\t${sha256(bytes)}\n`,
		);
	} else {
		links += 1;
		lines.push(`L\t${entry.rel}\t${mode(entry.stat)}\t${entry.target}\n`);
	}
}
const manifest = Buffer.from(lines.join(""));
const launcher = await readFile(LAUNCHER);
const framework = await readFile(FRAMEWORK);
const actual = {
	manifestBytes: manifest.byteLength,
	manifestHash: sha256(manifest),
	files,
	links,
	totalBytes,
	launcherHash: sha256(launcher),
	frameworkBytes: framework.byteLength,
	frameworkHash: sha256(framework),
};
for (const [key, value] of Object.entries(EXPECTED))
	if (actual[key] !== value) fail(`${key} ${actual[key]} != ${value}`);
process.stdout.write(
	`browser cohort valid: ${files} files, ${links} links, ${totalBytes} bytes\n`,
);

File: scripts/validate-browser-cohort.rb
require "digest"
require "find"

root = "/Users/pranav/Library/Caches/ms-playwright/chromium-1234/chrome-mac-arm64/Google Chrome for Testing.app"
launcher = File.join(root, "Contents/MacOS/Google Chrome for Testing")
framework = File.join(root, "Contents/Frameworks/Google Chrome for Testing Framework.framework/Versions/151.0.7922.34/Google Chrome for Testing Framework")
expected = {
  manifest_bytes: 62_239,
  manifest_hash: "25995bc88bf20b6de47b46eb3571b250989846a797c0b8924b8794627b6175fc",
  files: 326, links: 5, total_bytes: 372_002_382,
  launcher_hash: "a596b1cfc6353e987fcec8d71a23a28cd6a9e7a6b4e20b908e4c4fcffe51158e",
  framework_bytes: 237_813_488,
  framework_hash: "269114cf695f1c50b54e0816a1442e41dc468d28672e2dedc2036105fb5a8dbe"
}
entries = []
Find.find(root) do |path|
  next if path == root
  stat = File.lstat(path)
  rel = path.delete_prefix(root + "/")
  abort "browser cohort violation: unsafe path" if rel.match?(/[\t\n\r]/) || rel.split("/").include?("..")
  if stat.file?
    entries << ["F", rel, path, stat]
  elsif stat.symlink?
    target = File.readlink(path)
    abort "browser cohort violation: unsafe link" if target.match?(/[\t\n\r]/) || target.split("/").include?("..")
    entries << ["L", rel, target, stat]
  elsif !stat.directory?
    abort "browser cohort violation: unsupported object #{rel}"
  end
end
entries.sort_by! { |entry| entry[1].b }
files = 0; links = 0; total = 0
manifest = entries.map do |kind, rel, value, stat|
  mode = format("%04o", stat.mode & 0o7777)
  if kind == "F"
    bytes = File.binread(value); files += 1; total += bytes.bytesize
    "F\t#{rel}\t#{mode}\t#{bytes.bytesize}\t#{Digest::SHA256.hexdigest(bytes)}\n"
  else
    links += 1; "L\t#{rel}\t#{mode}\t#{value}\n"
  end
end.join.b
actual = {
  manifest_bytes: manifest.bytesize, manifest_hash: Digest::SHA256.hexdigest(manifest),
  files: files, links: links, total_bytes: total,
  launcher_hash: Digest::SHA256.file(launcher).hexdigest,
  framework_bytes: File.size(framework), framework_hash: Digest::SHA256.file(framework).hexdigest
}
expected.each { |key, value| abort "browser cohort violation: #{key} #{actual[key]} != #{value}" unless actual[key] == value }
puts "browser cohort valid: #{files} files, #{links} links, #{total} bytes"
